other/deny-default-service-account-bindings/deny-default-service-account-bindings.yaml

```diff
@@ -0,0 +1,34 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: deny-default-service-account-bindings
+  annotations:
+    policies.kyverno.io/title: Deny binding of default service accounts
+    policies.kyverno.io/category: Other
+    policies.kyverno.io/subject: Pod
+    kyverno.io/kyverno-version: 1.11.0
+    policies.kyverno.io/minversion: 1.10.0
+    kyverno.io/kubernetes-version: "1.26"
+    policies.kyverno.io/description: >-
+      The default service account is automatically mounted into all pods in a namespace unless 
+      explicitly overridden. If this account is bound to a Role or ClusterRole that grants extensive 
+      permissions, every pod in the namespace using the default service account will inherit these 
+      permissions. This setup can lead to unnecessary security risks if a pod is compromised, as an 
+      attacker could potentially gain access to other resources within the cluster.For an enhnaced 
+      security, using the default service account in RoleBindings is not recommended.
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+    - name: deny-default-service-account-bindings
+      match:
+        resources:
+          kinds:
+            - RoleBinding
+            - ClusterRoleBinding
+      validate:
+        message: "Using the default service account in RoleBindings is not allowed."
+        pattern:
+          subjects:
+            - kind: "ServiceAccount"
+              name: "!default"
```

Comment:
I haven't tested this, but my feeling is this may not work in all scenarios because `subjects[]` is an array of objects. If it isn't the first one in the list, it may bypass the policy. I suggest testing on your own to confirm. If confirmed, this would be better written as a deny rule.

Reply:
Hi @chipzoller,

With the change in policy, there's a new issue that the policy is excluding all the resources. Could you please point me to the issue here:

```
      validate:
        message: "Using the default service account in RoleBindings is not allowed."
        foreach:
          - list: "request.object.spec.subject[]"
            deny:
              conditions:
                all:
                  - key: "{{ element.kind }}"
                    operator: Equals
                    value: "ServiceAccount"
                  - key: "{{ element.name }}"
                    operator: Equals
                    value: "default"
```
 Following is the result:
 
```
 krishna@Krishnas-MacBook-Pro .tests % kyverno test .                                
Loading test  ( kyverno-test.yaml ) ...
  Loading values/variables ...
  Loading policies ...
  Loading resources ...
  Loading exceptions ...
  Applying 1 policy to 4 resources ...
  Checking results ...

│────│───────────────────────────────────────│───────────────────────────────────────│───────────────────────│────────│──────────│
│ ID │ POLICY                                │ RULE                                  │ RESOURCE              │ RESULT │ REASON   │
│────│───────────────────────────────────────│───────────────────────────────────────│───────────────────────│────────│──────────│
│ 1  │ deny-default-service-account-bindings │ deny-default-service-account-bindings │ RoleBinding/goodpod01 │ Pass   │ Excluded │
│ 2  │ deny-default-service-account-bindings │ deny-default-service-account-bindings │ RoleBinding/goodpod02 │ Pass   │ Excluded │
│ 3  │ deny-default-service-account-bindings │ deny-default-service-account-bindings │ RoleBinding/badpod01  │ Pass   │ Excluded │
│ 4  │ deny-default-service-account-bindings │ deny-default-service-account-bindings │ RoleBinding/badpod02  │ Pass   │ Excluded │
│────│───────────────────────────────────────│───────────────────────────────────────│───────────────────────│────────│──────────│


Test Summary: 4 tests passed and 0 tests failed
```

Reply:
With the following policy, we see the goodpods start, but badpods get skipped.
```
     validate:
        message: "Using the default service account in RoleBindings is not allowed."
        deny:
          conditions:
            all: 
            - key: "{{request.object.subject[].kind}}"
              operator: Equals
              value: "ServiceAccount"
            - key: "{{request.object.subject[].name}}"
              operator: Equals
              value: "default"
```

```
krishna@Krishnas-MacBook-Pro .kyverno-test % kyverno test .                           
Loading test  ( kyverno-test.yaml ) ...
  Loading values/variables ...
  Loading policies ...
  Loading resources ...
  Loading exceptions ...
  Applying 1 policy to 4 resources ...
  Checking results ...

│────│───────────────────────────────│───────────────────────────────│───────────────│────────│─────────────────────│
│ ID │ POLICY                        │ RULE                          │ RESOURCE      │ RESULT │ REASON              │
│────│───────────────────────────────│───────────────────────────────│───────────────│────────│─────────────────────│
│ 1  │ deny-default-service-accounts │ deny-default-service-accounts │ Pod/goodpod01 │ Pass   │ Ok                  │
│ 2  │ deny-default-service-accounts │ deny-default-service-accounts │ Pod/goodpod02 │ Pass   │ Ok                  │
│ 3  │ deny-default-service-accounts │ deny-default-service-accounts │ Pod/badpod01  │ Pass   │ Want fail, got skip │
│ 4  │ deny-default-service-accounts │ deny-default-service-accounts │ Pod/badpod02  │ Pass   │ Want fail, got skip │
│────│───────────────────────────────│───────────────────────────────│───────────────│────────│─────────────────────│
```